### PR TITLE
#1557 String-to-enum and enum-to-String

### PR DESCRIPTION
**Pull request by sjaakd**
working principle all source mappings must be mapped.

name based mapping is the default for all

**enum to enum**

1. ALL_REMAINING means, continue with name based (same name in target and source), what remains will be mapped to what is defined as ALL_REMAINING (default in case)
2. ALL_UNMAPPED means, stops after defined mapping, skips name based, map what remains to ALL_UNMAPPED (default in case)

**enum to string**, 

1. all not defined mappings will result in the source enum constant value being mapped to the string
2. ALL_REMAINING does not make sense there is no sense. We could claim all constants in source are present in the target, but then there would not be a default remaining. Or we could claim no constants in source are by default present, but then ALL_REMAINING would be the same as ALL_UNMAPPED. Better to avoid such ambiguous situation and give error.
3. ALL_UNMAPPED stops after handling defined mapping and proceeds to the default clause value.

**string to enum**

1. all not defined mappings will result in the target enum constant mapped from the string value
2. ALL_REMAINING does not make any sense on the same premise as above.
3. ALL_UNMAPPED stops after handling defined mapping and proceeds to the default clause value.

### CODE REVIEW

File: processor/src/test/java/org/mapstruct/ap/test/value/string2enum/StringToEnumMappingTest.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+@IssueKey( "1557" )
+@WithClasses({ OrderType.class, OrderMapper.class })
+@RunWith(AnnotationProcessorTestRunner.class)
+public class StringToEnumMappingTest {
+
+    @Test
+    public void testTheNormalStuff() {
```
Comment (filiphr) [inline processor/src/test/java/org/mapstruct/ap/test/value/string2enum/StringToEnumMappingTest.java]: I think that for this we can also test that the exception is thrown for an unknown `String`

Comment (sjaakd) [reply processor/src/test/java/org/mapstruct/ap/test/value/string2enum/StringToEnumMappingTest.java]: Hmm.. this is actually an interesting case. In the past we added this one for the possibility that a library containing the enum is updated runtime (and a new enum value has been added), but not yet compile time.

But what should we do for when the input is a `String`? I guess an `IllegalArgument` is justified. Or perhaps we should demand an `ANY_UNMAPPED` and even give an error when it's absent.. ?

Comment (filiphr) [reply processor/src/test/java/org/mapstruct/ap/test/value/string2enum/StringToEnumMappingTest.java]: I would say it is up to the user to define what they want. Throwing the exception is completely OK.

Comment (sjaakd) [reply processor/src/test/java/org/mapstruct/ap/test/value/string2enum/StringToEnumMappingTest.java]: > But what should we do for when the input is a `String`? I guess an `IllegalArgument` is justified. Or perhaps we should demand an `ANY_UNMAPPED` and even give an error when it's absent.. ?

thinking more about this.. This is also not the way forward. `ANY_UNMAPPED` stops after the defined mappings. Making it mandatory means the user always have to specify the mappings. 

Back to `ANY_REMAINING` and `ANY_UNMAPPED` for the string-to-enum case. And making one of these mandatory in this scenario.

Comment (sjaakd) [reply processor/src/test/java/org/mapstruct/ap/test/value/string2enum/StringToEnumMappingTest.java]: done.

**Comment by sjaakd** [comment]
@filiphr . thanks for reviewing.. I changed the behaviour for string to enum mapping. I don't think our code should throw `illegalArgument` if we can resolve things by making the api more consistent. So, that's what I did.

Have a look and see if you agree.

File: processor/src/test/java/org/mapstruct/ap/test/value/string2enum/StringToEnumMappingTest.java (hunk 1 above)
Comment (sjaakd) [reply processor/src/test/java/org/mapstruct/ap/test/value/string2enum/StringToEnumMappingTest.java]: > I would say it is up to the user to define what they want. Throwing the exception is completely OK.

Generating an `IllegalArgumentException` for this case might surprise the user when running the code. A String not matching the enum constant will be pretty common. 

I prefer to make the user aware that he should make a choice here at the expense of a little bit more verbosity.

Comment (filiphr) [reply processor/src/test/java/org/mapstruct/ap/test/value/string2enum/StringToEnumMappingTest.java]: I would say that having a compile error is too strict. Perhaps only a warning so they are aware about it. Perhaps they want to do it on purpose like that.

File: documentation/src/main/asciidoc/chapter-8-mapping-values.asciidoc [hunk 2]
```diff
@@ -67,7 +67,7 @@ public class OrderMapperImpl implements OrderMapper {
 By default an error will be raised by MapStruct in case a constant of the source enum type does not have a corresponding constant with the same name in the target type and also is not mapped to another constant via `@ValueMapping`. This ensures that all constants are mapped in a safe and predictable manner. The generated
 mapping method will throw an IllegalStateException if for some reason an unrecognized source value occurs.
 
-MapStruct also has a mechanism for mapping any remaining (unspecified) mappings to a default. This can be used only once in a set of value mappings. It comes in two flavors: `<ANY_REMAINING>` and `<ANY_UNMAPPED>`.
+MapStruct also has a mechanism for mapping any remaining (unspecified) mappings to a default. This can be used only once in a set of value mappings and only applies to the source. It comes in two flavors: `<ANY_REMAINING>` and `<ANY_UNMAPPED>`. The cannot be used at the same time.
```
[the remark this answers is not in the document]
Comment (sjaakd) [reply documentation/src/main/asciidoc/chapter-8-mapping-values.asciidoc]: done.

File: documentation/src/main/asciidoc/chapter-8-mapping-values.asciidoc [hunk 3]
```diff
@@ -1,7 +1,7 @@
 [[mapping-enum-types]]
 == Mapping Values
 
-=== Mapping enum types
+=== Mapping enum to enum types
```
[the remark this answers is not in the document]
Comment (sjaakd) [reply documentation/src/main/asciidoc/chapter-8-mapping-values.asciidoc]: added a link name

**Comment by sjaakd** [comment]
@filiphr thanks for reviewing.

**Pull request merged**